### ISSUE TITLE
Train models on Multi GPU

### ISSUE THREAD

**Issue by christianlancho** [nicolas-chaulet/torch-points3d#347]
Hi,

First of all thanks for the great library. I'm trying to train a model, specifically pointnet2 charlesmsg, and in order to use a higher batch size I would want to train the model in two GPUs. I've wrapped the atributte `model` of the class `PointNet2_D` into a `DataParallel`. The main problem is that the model is loaded in both GPUs but before the training starts there's an error:

`Expected tensor for argument #1 'input' to have the same device as tensor for argument #2 'weight'`

Basically, the second GPU has loaded the model but not the data. Searching on internet I found someone with the same problem ( https://github.com/pytorch/pytorch/issues/8637 ) and the main cause It might be using in the method  `forward` methods save as attribute .In summary,  ¿Is there a way implemented in the code to use multi GPU training easily?

**Comment by nicolas-chaulet** [comment]
Not sure about that to be honnest... Looks like you should change the forward method a bit so that it takes the input data directly (as opposed to using set_input). You could for example change it just like here:
https://github.com/nicolas-chaulet/torch-points3d/blob/c939c995cc01857a0c797c7d5cef147f1b15d1a6/torch_points3d/applications/pointnet2.py#L168

Then modify the set_input method, just add `device = self.device` here: https://github.com/nicolas-chaulet/torch-points3d/blob/c939c995cc01857a0c797c7d5cef147f1b15d1a6/torch_points3d/models/segmentation/pointnet2.py#L74 

Finally, remove the set_input calls in the trainer and call the forward method on the data directly. Let me know if that works! If it does then we would welcome a PR it's been on our roadmap for some time now

**Comment by christianlancho** [comment]
Thanks for your reply! 

I have tryed out what you suggest but I get the same error. Because of this I have done some tests with a simple multi GPU example from the pytorch documentation. In the toy example, if you use two GPUs the forward method is called twice, once per GPU. However, in this code, the forward method is called only once even if I wrap the model into a DataParallel. The model is loaded on both GPU but not the data. 

I also tryed to wrap the model in different levels, nonetheless, the only place where two GPUs are used (at least loading the model) is if you wrap the attribute `model` of the class `PointNet2_D` into a DataParallel.

**Comment by nicolas-chaulet** [comment]
Hi! Do you have some code that I could take a look at?

**Comment by christianlancho** [comment]
> Hi! Do you have some code that I could take a look at?
> …

Yes, of course, ¡Thanks for your interest! I have updated the code so I'm going to explain the few changes I've done. 
- I've wrapped the model into a DataParallel adding this code 

`self._model.model = torch.nn.DataParallel(self._model.model, device_ids=[2, 3])`  before this line
https://github.com/nicolas-chaulet/torch-points3d/blob/46a3b622b555e720b68acce62a123f95a36026da/torch_points3d/trainer.py#L120

- I'm going to call `set_input()` inside the forward method, so I've commented the line 185 and add the param `data` to the method `optimize_parameters()`
https://github.com/nicolas-chaulet/torch-points3d/blob/46a3b622b555e720b68acce62a123f95a36026da/torch_points3d/trainer.py#L185-L186

- In the optimize_parameters() I've changed line 196 into `self.forward(epoch=epoch, data=data)`

https://github.com/nicolas-chaulet/torch-points3d/blob/46a3b622b555e720b68acce62a123f95a36026da/torch_points3d/models/base_model.py#L196

- I changed slightly the set_input method removing attribute `device` and adding the line that you told me last time `device = self.device` after this line 

https://github.com/nicolas-chaulet/torch-points3d/blob/46a3b622b555e720b68acce62a123f95a36026da/torch_points3d/models/segmentation/pointnet2.py#L74

- Finally, I change `forward()` method adding this line `self.set_input(kwargs['data'])` before this

https://github.com/nicolas-chaulet/torch-points3d/blob/46a3b622b555e720b68acce62a123f95a36026da/torch_points3d/models/segmentation/pointnet2.py#L96

I get the same error:

`RuntimeError: Expected tensor for argument #1 'input' to have the same device as tensor for argument #2 'weight'; but device 2 does not equal 3 (while checking arguments for cudnn_convolution)`

**Comment by nicolas-chaulet** [comment]
Could you push your code to a fork? It will be easier like that...

**Comment by christianlancho** [comment]
> Could you push your code to a fork? It will be easier like that...

Yeah of course, there aren't too many changes so I thought maybe this way was easier but you are probably right. I have pushed my code into a fork:

https://github.com/christianlancho/torch-points3d.git

Thanks.

**Comment by nicolas-chaulet** [comment]
Yes looks good indeed. Not sure what's wrong, could you print the device of "data"? If I understand data parallel right the data should be put on the device automatically. Not sure though...

**Comment by christianlancho** [comment]
> Yes looks good indeed. Not sure what's wrong, could you print the device of "data"? If I understand data parallel right the data should be put on the device automatically. Not sure though...

Thanks for your time,

The main problema I see is that the forward method is only called once. The behaviour that I expected was to call `forward()` with data in device 2 and, at the same time, call `forward()` with data in device 3. That what happens in the toy example I have tested. However, when I put  a print() in the `forward()` to check the device, only print once, and the value is the device used by default, in my case 2. Due to this, and taking into account that the `DataParallel()` should do this automatically, I thought that the problem could be how the `forward()` mehod works (using methods and attibutes from other classes). I say this because of the post I found (I mention it in the first issue), but no idea.

**Comment by chaitjo** [comment]
Hi, does the library support multi-GPU training at the moment?

**Comment by nicolas-chaulet** [comment]
No, not yet. There is quite a bit of work to do in order to get that working unfortunatelly...
